src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyDiscoveryManager.cs

```diff
@@ -182,21 +185,45 @@ public bool HandlePartialDiscoveryComplete(IProxyDiscoveryManager proxyDiscovery
 
     private List<ProviderSpecificWorkload<DiscoveryCriteria>> SplitToWorkloads(DiscoveryCriteria discoveryCriteria, Dictionary<string, TestRuntimeProviderInfo> sourceToTestHostProviderMap)
     {
+        var sources = discoveryCriteria.Sources;
+        // Each source is grouped with its respective provider.
+        var providerGroups = sources
+            .Select(source => new ProviderSpecificWorkload<string>(source, sourceToTestHostProviderMap[source]))
+            .GroupBy(psw => psw.Provider);
+
         List<ProviderSpecificWorkload<DiscoveryCriteria>> workloads = new();
-        foreach (var source in discoveryCriteria.Sources)
+        foreach (var group in providerGroups)
         {
-            var testHostProviderInfo = sourceToTestHostProviderMap[source];
-            var runsettingsXml = testHostProviderInfo.RunSettings;
-            var updatedDiscoveryCriteria = new ProviderSpecificWorkload<DiscoveryCriteria>(NewDiscoveryCriteriaFromSourceAndSettings(source, discoveryCriteria, runsettingsXml), testHostProviderInfo);
-            workloads.Add(updatedDiscoveryCriteria);
+            var testhostProviderInfo = group.Key;
+            // If the run is not parallel and the host is shared, put all testcases on single testhost.
+            if (!_isParallel && testhostProviderInfo.Shared)
+            {
+                var runsettings = testhostProviderInfo.RunSettings;
+                var sourcesToDiscover = group.Select(w => w.Work).ToArray();
+                var updatedCriteria = NewDiscoveryCriteriaFromSourceAndSettings(sourcesToDiscover, discoveryCriteria, runsettings);
+                var workload = new ProviderSpecificWorkload<DiscoveryCriteria>(updatedCriteria, testhostProviderInfo);
+                workloads.Add(workload);
+            }
+            else
+            {
+                // Create one workload for each source
+                foreach (var w in group.ToList())
+                {
+                    var runsettings = testhostProviderInfo.RunSettings;
+                    var sourcesToDiscover = new[] { w.Work };
+                    var updatedCriteria = NewDiscoveryCriteriaFromSourceAndSettings(sourcesToDiscover, discoveryCriteria, runsettings);
+                    var workload = new ProviderSpecificWorkload<DiscoveryCriteria>(updatedCriteria, testhostProviderInfo);
+                    workloads.Add(workload);
```

Comment:
Outside for the sources to discover the logic of the loop is exactly the same as in previous if. I think it might be worth to update the local static function to either handle `ProviderSpecificWorkload` creation and locally do `workloads.Add(workload)` or also do this from the function.

Reply:
I end up grouping the sources into a collection of arrays. 

It re-uses the logic because we then run on collection of string arrays. It it still not very elagant, and you will hate me for that type. But it is clearer than using IEnumerable<IEnumerable<string>> 😁